### PR TITLE
[Feature] Support dynamic speculative decoding with DFlash and Unified dynamic selection logic

### PR DESCRIPTION
**Pull request by StanislavII**
## Summary
Follow up from https://github.com/vllm-project/vllm-ascend/pull/13216
This PR unifies the dynamic speculative decoding logic for **DFlash** and **DSpark**.

Both methods now use the same dynamic verification pipeline:

```text
token confidence
    -> token acceptance probabilities
    -> cumulative survival probabilities
    -> shared verify budget
    -> per-request verify lengths
```

The shared scheduler handles:

- `budget_threshold`
- `budget_update_interval`
- `initial_verify_budget_per_req`
- cumulative survival computation
- global verification-budget update
- per-request budget allocation
- dynamic-spec buffers and state

The only method-specific part is how token confidence is computed.

## DFlash confidence

DFlash does not have a dedicated confidence head, so the softmax probability of the selected greedy token is used as its confidence estimate.

For logits `z_i`:

```text
t_i = argmax_j z_{i,j}
```

and the confidence of the selected token is:

```text
p_i = exp(z_{i,t_i}) / sum_j exp(z_{i,j})
```

This is computed directly with the fused max-softmax kernel without materializing the full softmax tensor.

## DSpark confidence

DSpark uses its learned confidence head:

```text
p_i = sigmoid(c_i)
```

where `c_i` is the confidence-head output for speculative position `i`.

## Unified survival-based scheduling

After obtaining per-token probabilities, both DFlash and DSpark use exactly the same logic.

Prefix survival probability is estimated as:

```text
S_k = product_{i=1..k} p_i
```

In code:

```python
survival = torch.cumprod(
    token_probs,
    dim=1,
)
```

The survival probabilities are then used both to update the shared verification budget and to distribute that budget across requests.

Requests with higher predicted prefix survival receive longer verification lengths, while lower-confidence requests receive shorter ones.
## Tests
**Target model**: Qwen/Qwen3-8B
**DFlash**: z-lab/Qwen3-8B-DFlash-b16
**DSpark**: deepseek-ai/dspark_qwen3_8b_block7
**Confidence scores computation**: DFlash - the softmax probability of the selected greedy, DSpark - confidence head prediction

**NB:** current threshold is not optimized for the specific draft model and was set as **budget_threshold = 0.3** for both DFlash and DSpark. Lower threshold will reduce verification budget and pressure but will also reduce the acceptance length, so for the better performance this parameter should be optimized for corresponding target/draft model and data-flow, the logic values for this is into [0.2-0.4]
**All tests are conducted on bs = 128**

Method | Draft | EAGER tok/s OFF → ON | Δ | FULL tok/s OFF → ON | Δ | Acceptance OFF → ON | Δ 
-- | -- | -- | -- | -- | -- | -- | --
Target | — | 1517 | — | 2454 | — | — | —
DFlash | 9 | 2475 → 2904 | +17.33% | 2649 → 3213 | +21.29% | 3.46 → 3.25 | -0.21
  | 7 | 2612 → 2920 | +11.79% | 2788 → 3165 | +13.52% | 3.32 → 3.17 | -0.15
  | 5 | 2804 → 2871 | +2.39% | 3051 → 3182 | +4.29% | 3.04 → 2.94 | -0.10
DSpark | 9 | 2812 → 3189 | +13.41% | 3133 → 3551 | +13.34% | 4.27 → 4.09 | -0.18
  | 7 | 3022 → 3238 | +7.15% | 3428 → 3704 | +8.05% | 4.16 → 4.06 | -0.10
  | 5 | 3149 → 3092 | -1.81% | 3639 → 3593 | -1.26% | 3.70 → 3.66 | -0.04

Using the average over K = 5, 7, 9, and normalizing each graph mode to its own Target throughput:

Graph mode | Target | DFlash | + Dynamic | DSpark | + Dynamic
-- | -- | -- | -- | -- | --
EAGER | ×1.000 | ×1.734 | ×1.911 **(x1.10)** | ×1.974 | ×2.092 **(x1.06)**
FULL | ×1.000 | ×1.153 | ×1.299 **(x1.13)** | ×1.385 | ×1.474 **(x1.06)**

Extreme Case for Dflash
Graph mode | DFlash (15) | Dynamic DFlash (15)
-- | -- | -- 
FULL| 1991 tok/sec | 3063 tok/sec (x1.54)

**GLM5.2w4a8 + Dspark: TP8EP bs =128, max_tokens = 512**

Due to CUDAGraphMode.FULL is not supported with DSA and SFA Backends and setting FULL_DECODE_ONLY is not convenient for Dynamic tests (Not uniform batch -> Eager mode), next test are conducted on PIECEWISE

Method | Draft | PIECEWISE tok/s OFF → ON | Δ | Acceptance OFF → ON | Δ 
-- | -- | -- | -- | -- | -- 
Dspark| 9 | 730.2 → 1115.3 | +52.73% | 3.02 → 2.91 | -0.11
  | 7 | 819.1 → 962.9 | +17.56% | 2.96 → 2.84 | -0.12
  | 5 | 921.6 → 902.4 | -2.1% | 2.82 → 2.78 | -0.04

**DeepSeek-V4-Flashw4a8 + Dspark: TP4EP bs =128, max_tokens = 1024**
Is not supported yet for the current main, need new PR enabling conf_head for DS (https://github.com/vllm-project/vllm-ascend/pull/14249)
Method | Draft | PIECEWISE tok/s OFF → ON | Δ | Acceptance OFF → ON | Δ 
-- | -- | -- | -- | -- | -- 
Dspark| 9 | 1102 → 1320 | +19.78% | 3.88 → 3.72 | -0.16
  | 7 | 1021 → 1210 | +18.62% | 3.88 → 3.79 | -0.09
  | 5 | 1023 → 1019 | -0.4%| 3.55 → 3.51 | -0.04
## Refactoring

Dynamic-spec state and buffers are moved from the individual proposer implementations into a shared scheduler.

As a result, proposer-specific code is reduced to confidence computation, while budget calculation and token allocation are shared between DFlash and DSpark.

- vLLM version: v0.26.0
- vLLM main: https://github.com/vllm-project/vllm/commit/58d3918e3ea0a544ffedadad2ba84559e9c51d8f

### CODE REVIEW

**Review by Tflowers-0129 — APPROVED** [review]

File: vllm_ascend/spec_decode/dspark_proposer.py [hunk 1]
```diff
@@ -65,40 +65,17 @@ def __init__(
...
-            self.max_batch_size,
-            dtype=torch.int32,
-            device=device,
-        )
-        self._keep_lens = torch.zeros(
-            (self.max_batch_size,),
-            dtype=torch.int32,
-            device=self.device,
-        )
+        self.dynamic_spec = None
```
Comment (drslark) [inline vllm_ascend/spec_decode/dspark_proposer.py:69]: Good.

**Review by drslark — COMMENTED** [review]
LGTM.

**Review by drslark — APPROVED** [review]
Thanks for your efforts.

**Review by zzzzwwjj — APPROVED** [review]
